services/fdb/server/fdbmovedata.go

```diff
@@ -0,0 +1,183 @@
+	"io"
+	"math/rand"
+	"os/exec"
+	"sync"
+
+	"github.com/Snowflake-Labs/sansshell/services"
+	pb "github.com/Snowflake-Labs/sansshell/services/fdb"
+	"github.com/go-logr/logr"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	// fdb_move_orchestrator binary location
+	// Will be checked into:
+	// https://github.com/apple/foundationdb/tree/snowflake/release-71.3/bindings/python/fdb/fdb_move_orchestrator.py
+	FDBMoveOrchestrator     string
+	generateFDBMoveDataArgs = generateFDBMoveDataArgsImpl
+)
+
+type fdbmovedata struct {
+	mu     sync.Mutex
+	id     int64
+	cmd    *exec.Cmd
+	stdout io.ReadCloser
+	stderr io.ReadCloser
+}
+
+func generateFDBMoveDataArgsImpl(req *pb.FDBMoveDataCopyRequest) ([]string, error) {
+	command := []string{FDBMoveOrchestrator}
+
+	var args []string
+	args = append(args, "--cluster", req.ClusterFile)
+	args = append(args, "--tenant-group", req.TenantGroup)
+	args = append(args, "--src-name", req.SourceCluster)
+	args = append(args, "--dst-name", req.DestinationCluster)
+	args = append(args, "--num-procs", fmt.Sprintf("%d", req.NumProcs))
+
+	command = append(command, args...)
+	return command, nil
+}
+
+func (s *fdbmovedata) FDBMoveDataCopy(ctx context.Context, req *pb.FDBMoveDataCopyRequest) (*pb.FDBMoveDataCopyResponse, error) {
+	lockSuccess := s.mu.TryLock()
+	if !(lockSuccess) {
+		return nil, status.Errorf(codes.Internal, "Copy or Wait command already running")
+	}
+	defer s.mu.Unlock()
+	logger := logr.FromContextOrDiscard(ctx)
+	// The sansshell server should only run one copy command at a time
+	if !(s.cmd == nil) {
+		logger.Info("existing command already running. returning early")
+		earlyresp := &pb.FDBMoveDataCopyResponse{
+			Id:       s.id,
+			Existing: true,
+		}
+		return earlyresp, nil
+	}
+	s.id = rand.Int63()
+	command, err := generateFDBMoveDataArgs(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Add env vars for python and python bindings
+	cmd := exec.Command(command[0], command[1:]...)
+	cmd.Env = append(cmd.Environ(), "PYTHONPATH=/opt/rh/rh-python36/root/lib/python3.6/site-packages/")
+	cmd.Env = append(cmd.Environ(), "PATH=/opt/rh/rh-python36/root/bin")
+	cmd.Env = append(cmd.Environ(), "PYTHONUNBUFFERED=true")
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("executing local command", "cmd", cmd.String())
+	s.cmd = cmd
+	s.stdout = stdout
+	s.stderr = stderr
+	err = s.cmd.Start()
+	if err != nil {
+		s.cmd = nil
+		s.id = 0
+		return nil, status.Errorf(codes.Internal, "error running fdbmovedata cmd (%+v): %v", command, err)
+	}
+
+	resp := &pb.FDBMoveDataCopyResponse{
+		Id:       s.id,
+		Existing: false,
+	}
+	return resp, nil
+}
+
+func (s *fdbmovedata) FDBMoveDataWait(req *pb.FDBMoveDataWaitRequest, stream pb.FDBMove_FDBMoveDataWaitServer) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !(req.Id == s.id) {
```

Comment:
probably a good idea to add some logging in this function. It'd be helpful for debugging

Reply:
How do you create a logger without `logger := logr.FromContextOrDiscard(ctx)`? This function does not have `ctx`

Reply:
You can get context from stream
```
ctx := stream.Context()
logger := logr.FromContextOrDiscard(ctx)
```